pyquil/paulis.py

```diff
@@ -203,7 +203,7 @@ def _multiply_factor(self, factor, index):
 
         return new_term
 
-    def __mul__(self, term):
+    def __mul__(self, term: Union['PauliTerm', 'PauliSum', 'Number']) -> 'PauliTerm':
```

Comment:
I am basing the return type annotation on the docstring. But this should in general be a `Union['PauliTerm', 'PauliSum']`, because sometimes it does return a `PauliSum`.

Reply:
In that case, I would vote update the docstring and have the type hint reflect the actual return type.

Or better yet, just remove all these `:rtype:` lines from the docstrings for any methods that have a return type hint. We use a sphinx plugin to automatically add the return type if a type hint is present, so they are now redundant.

Reply:
Updated the return type. For removing `:rtype:`, I think it should be done in a separate PR, to keep things modular.

Reply:
It's not a blocker, but unless you're planning to file an immediate follow-on PR, my preference would be to handle it in this PR. Otherwise, these things tend to slip through the cracks and eventually the docstring types become stale or out-of-sync with the type hints.